### PR TITLE
Outgoing

### PR DESCRIPTION
**Pull request by yarinBenisty**

### CODE REVIEW

File: src/express/outgoing/manager.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+export const getOutgoingFileById = async (userId: string, fileId: mongoose.Types.ObjectId): Promise<IOutgoingFile> => {
+    const response = await transfersCrud
+        .get(`/${userId}/outgoing/${fileId}`)
+        .catch(serviceErrorHandler('Failed to get file'));
+    return response.data;
+};
+
+export const getOutgoingFiles = async (filters: IOutgoingFileFilters): Promise<IOutgoingFile[]> => {
+    const response = await transfersCrud
+        .get(`/${filters.from}/outgoing`, { params: filters })
```
Comment (MichaelSimkin) [inline src/express/outgoing/manager.ts:59]: make userId function parameter, (changes in controller as well)
```suggestion
        .get(`/${userId}/outgoing`, { params: { ...filters, from: userId } })
```

File: src/express/outgoing/controller.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+export const createOutgoingFile = async (req: Request<{}, any, {}, any>, res: Response) => {
+    res.json(await OutgoingManager.createOutgoingFile(req.user.id, req, req.query));
+};
+
+export const getOutgoingFile = async (
+    req: Request<{ userId: string; fileId: mongoose.Types.ObjectId }>,
+    res: Response,
+) => {
+    res.json(await OutgoingManager.getOutgoingFileById(req.params.userId, req.params.fileId));
```
Comment (MichaelSimkin) [inline src/express/outgoing/controller.ts:13]: we get user from req.user.id not from params in all routes
```suggestion
    res.json(await OutgoingManager.getOutgoingFileById(req.user.id, req.params.fileId));
```

File: src/express/outgoing/router.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import { Router } from 'express';
+import wrapMiddleware from '../../utils/express';
+import ValidateRequest from '../../utils/joi';
+import * as outgoingController from './controller';
+import * as outgoingValidator from './validator.schema';
+
+const outgoingRouter: Router = Router();
+
+outgoingRouter.post(
+    '/:userId/outgoing',
```
Comment (MichaelSimkin) [inline src/express/outgoing/router.ts:10]: remove :userId in all routes
```suggestion
    '/outgoing',
```

File: src/express/outgoing/validator.schema.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    },
+    params: {
+        userId: Joi.string().required(),
+    },
+});
+
+export const getOutgoingFileRequestSchema = Joi.object({
+    params: {
+        fileId: JoiObjectId.required(),
+        userId: Joi.string().required(),
```
Comment (MichaelSimkin) [inline src/express/outgoing/validator.schema.ts:28]: remove in all validators userId
```suggestion

```

File: src/express/router.ts [hunk 5]
```diff
@@ -1,16 +1,22 @@
...
-/* SHRAGA AUTHENTICATION ROUTES */
-appRouter.get('/auth/login', shragaLoginMiddleware); // UI will redirect to this route to login with shraga
-appRouter.post('/auth/callback', shragaCallbackMiddleware); // Shraga will redirect to this route after login
+appRouter.get('/config', (_, res: Response, next: NextFunction) => {
+    res.json({});
+    next();
+});
+
+appRouter.use('/api/users', outgoingRouter);
+appRouter.use('/api/users', userRouter);
```
Comment (MichaelSimkin) [inline src/express/router.ts:19]: remove /users
```suggestion
appRouter.use('/api', outgoingRouter);
appRouter.use('/api', userRouter);
```

File: src/express/user/controller.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { Request, Response } from 'express';
+import * as UserManager from './manager';
+
+export const findPhonebookUsers = async (req: Request<{}, any, { users: string[] }, any>, res: Response) => {
+    res.json(await UserManager.checkPhonebookUsers(req.body.users));
```
Comment (MichaelSimkin) [inline src/express/user/controller.ts:5]:
```suggestion
    res.json(await UserManager.checkPhonebookUsers(req.query.query, req.query.limit));
```

File: src/express/user/controller.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+};
+
+export const findKartoffelUsers = async (req: Request<{}, any, {}, any>, res: Response) => {
+    res.json(await UserManager.checkKartoffelUsers(req.query.source, req.query));
```
Comment (MichaelSimkin) [inline src/express/user/controller.ts:9]:
```suggestion
    res.json(await UserManager.checkKartoffelUsers(req.query.query, req.query.source));
```

**Review by MichaelSimkin — CHANGES_REQUESTED** [review]

**Comment by MichaelSimkin** [comment]
Change folder `user` to `users`

File: src/express/outgoing/manager.ts (hunk 1 above)
Comment (yarinBenisty) [reply src/express/outgoing/manager.ts:59]: done

File: src/express/outgoing/controller.ts (hunk 2 above)
Comment (yarinBenisty) [reply src/express/outgoing/controller.ts:13]: done

File: src/express/outgoing/router.ts (hunk 3 above)
Comment (yarinBenisty) [reply src/express/outgoing/router.ts:10]: done

File: src/express/outgoing/validator.schema.ts (hunk 4 above)
Comment (yarinBenisty) [reply src/express/outgoing/validator.schema.ts:28]: done

File: src/express/router.ts (hunk 5 above)
Comment (yarinBenisty) [reply src/express/router.ts:19]: done

File: src/express/user/controller.ts (hunk 7 above)
Comment (yarinBenisty) [reply src/express/user/controller.ts:9]: done

**Comment by yarinBenisty** [comment]
folder name changed :D

**Comment by yarinBenisty** [comment]
dont forget to approve

File: src/express/outgoing/manager.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+};
+
+export const getOutgoingFiles = async (userId: string, filters: IOutgoingFileFilters): Promise<IOutgoingFile[]> => {
+    const response = await transfersCrud
+        .get(`/${userId}/outgoing`, { params: { ...filters, from: userId } })
+        .catch(serviceErrorHandler('Failed to get files'));
+    return response.data;
+};
+
+export const getOutgoingFileStatusById = async (fileId: mongoose.Types.ObjectId): Promise<IFileFlow> => {
```
Comment (MichaelSimkin) [inline src/express/outgoing/manager.ts:64]: Add userId verificatiion. Make axios request to get file where from=userId and check if exists.

File: src/express/router.ts [hunk 9]
```diff
@@ -1,16 +1,22 @@
...
+import { NextFunction, Response, Router } from 'express';
 import { StatusCodes } from 'http-status-codes';
 import { shragaCallbackMiddleware, shragaLoginMiddleware } from './auth';
 import { ServerError } from './error';
-import featuresRouter from './incomingFile/router';
+import outgoingRouter from './outgoing/router';
+import userRouter from './users/router';
 
-const appRouter = Router();
+const appRouter: Router = Router();
```
Comment (MichaelSimkin) [inline src/express/router.ts:8]: Type not needed. (infered from Router();)
```suggestion
const appRouter = Router();
```

File: src/utils/joi.ts [hunk 10]
```diff
@@ -1,8 +1,12 @@
 import { Request } from 'express';
 import * as Joi from 'joi';
+import mongoose from 'mongoose';
 import wrapMiddleware from './express';
 
-export const JoiMongoObjectId = Joi.string().hex().length(24);
+export const JoiObjectId = Joi.string()
+    .hex()
+    .length(24)
+    .custom((value) => new mongoose.Types.ObjectId(value));
```
Comment (MichaelSimkin) [inline src/utils/joi.ts:9]: We should work with strings in gateway

```suggestion
export const JoiMongoObjectId = Joi.string().hex().length(24);
```

File: src/express/outgoing/manager.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        serviceErrorHandler('Failed to send outgoing file to dropbox', async () => {
+            await transfersCrud.delete(`/${userId}/outgoing/${fileId}`);
+            await storageService.delete(`/bucket/${file.from}/object/${fileId}`);
+        }),
+    );
+
+    return file;
+};
+
+export const getOutgoingFileById = async (userId: string, fileId: mongoose.Types.ObjectId): Promise<IOutgoingFile> => {
```
Comment (MichaelSimkin) [inline src/express/outgoing/manager.ts:50]:
```suggestion
export const getOutgoingFileById = async (userId: string, fileId: string): Promise<IOutgoingFile> => {
```

File: src/express/outgoing/manager.ts (hunk 8 above)
Comment (MichaelSimkin) [inline src/express/outgoing/manager.ts:64]:
```suggestion
export const getOutgoingFileStatusById = async (userId: string, fileId: string): Promise<IFileFlow> => {
```

File: src/express/outgoing/manager.ts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+};
+
+export const getOutgoingFileStatusById = async (fileId: mongoose.Types.ObjectId): Promise<IFileFlow> => {
+    const response = await statusService
+        .get(`/fileFlow/files/statusId/${fileId}`)
+        .catch(serviceErrorHandler('Failed to get file status'));
+    return response.data;
+};
+
+export const deleteOutgoingFile = async (userId: string, fileId: mongoose.Types.ObjectId): Promise<IOutgoingFile> => {
```
Comment (MichaelSimkin) [inline src/express/outgoing/manager.ts:71]:
```suggestion
export const deleteOutgoingFile = async (userId: string, fileId: string): Promise<IOutgoingFile> => {
```

File: src/express/outgoing/interface.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
+import * as mongoose from 'mongoose';
+import { externalNetwork } from '../interface';
+
+export interface IOutgoingFile {
+    size: number;
+    _id: mongoose.Types.ObjectId;
```
Comment (MichaelSimkin) [inline src/express/outgoing/interface.ts:6]:
```suggestion
    _id: string;
```

File: src/express/outgoing/interface.ts [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    classification: string;
+    info: string;
+    destination: externalNetwork;
+    createdAt: Date;
+}
+
+export type INewOutgoingFile = Omit<IOutgoingFile, '_id' | 'createdAt' | 'from'>;
+
+export type IOutgoingFileFilters = {
+    _id?: mongoose.Types.ObjectId;
```
Comment (MichaelSimkin) [inline src/express/outgoing/interface.ts:20]:
```suggestion
    _id?: string;
```

**Review by MichaelSimkin — CHANGES_REQUESTED** [review]
Small changes needed

File: src/express/outgoing/controller.ts [hunk 15]
```diff
@@ -0,0 +1,29 @@
+import { Request, Response } from 'express';
+import mongoose from 'mongoose';
+import * as OutgoingManager from './manager';
+
+export const createOutgoingFile = async (req: Request<{}, any, {}, any>, res: Response) => {
+    res.json(await OutgoingManager.createOutgoingFile(req.user.id, req, req.query));
+};
+
+export const getOutgoingFile = async (
+    req: Request<{ userId: string; fileId: mongoose.Types.ObjectId }>,
```
[the remark this answers is not in the document]
Comment (yarinBenisty) [reply src/express/outgoing/controller.ts:10]: done

File: src/express/outgoing/controller.ts [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+) => {
+    res.json(await OutgoingManager.getOutgoingFileById(req.user.id, req.params.fileId));
+};
+
+export const getOutgoingFiles = async (req: Request, res: Response) => {
+    res.json(await OutgoingManager.getOutgoingFiles(req.user.id, req.query));
+};
+
+export const getOutgoingFileStatus = async (req: Request<{ fileId: mongoose.Types.ObjectId }>, res: Response) => {
+    res.json(await OutgoingManager.getOutgoingFileStatusById(req.params.fileId));
```
[the remark this answers is not in the document]
Comment (yarinBenisty) [reply src/express/outgoing/controller.ts:21]: done

File: src/express/outgoing/manager.ts (hunk 8 above)
Comment (yarinBenisty) [reply src/express/outgoing/manager.ts:64]: done

File: src/express/router.ts (hunk 9 above)
Comment (yarinBenisty) [reply src/express/router.ts:8]: done

File: src/utils/joi.ts (hunk 10 above)
Comment (yarinBenisty) [reply src/utils/joi.ts:9]: done

File: src/express/outgoing/manager.ts (hunk 11 above)
Comment (yarinBenisty) [reply src/express/outgoing/manager.ts:50]: done

File: src/express/outgoing/manager.ts (hunk 8 above)
Comment (yarinBenisty) [reply src/express/outgoing/manager.ts:64]: done

File: src/express/outgoing/manager.ts (hunk 12 above)
Comment (yarinBenisty) [reply src/express/outgoing/manager.ts:71]: done

File: src/express/outgoing/interface.ts (hunk 14 above)
Comment (yarinBenisty) [reply src/express/outgoing/interface.ts:20]: done

File: src/express/outgoing/interface.ts (hunk 13 above)
Comment (yarinBenisty) [reply src/express/outgoing/interface.ts:6]: done

**Review by sternomer — APPROVED** [review]

File: src/express/outgoing/manager.ts [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+import { checkPhonebookUsers } from '../users/manager';
+import { IFileFlow, INewOutgoingFile, IOutgoingFile, IOutgoingFileFilters } from './interface';
+
+export const createOutgoingFile = async (
+    userId: string,
+    req: Request,
+    newFile: INewOutgoingFile,
+): Promise<IOutgoingFile> => {
+    if (!checkPhonebookUsers(newFile.to))
+        throw new ServerError(StatusCodes.INTERNAL_SERVER_ERROR, 'Failed to get users.');
```
Comment (MichaelSimkin) [inline src/express/outgoing/manager.ts:16]:
```suggestion
        throw new ServerError(StatusCodes.BAD_REQUEST, 'Failed to find users.');
```

File: src/express/outgoing/manager.ts [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        .catch(serviceErrorHandler('Failed to get files'));
+    return response.data;
+};
+
+export const getOutgoingFileStatusById = async (userId: string, fileId: string): Promise<IFileFlow> => {
+    const file = (
+        await transfersCrud.get(`/${userId}/outgoing/${fileId}`).catch(serviceErrorHandler('Failed to get file'))
+    ).data;
+
+    if (file.from !== userId) throw new ServerError(StatusCodes.INTERNAL_SERVER_ERROR, 'User does not own file');
```
Comment (MichaelSimkin) [inline src/express/outgoing/manager.ts:68]:
```suggestion
    if (file.from !== userId) throw new ServerError(StatusCodes.FORBIDDEN, 'User does not own file');
```

**Review by MichaelSimkin — CHANGES_REQUESTED** [review]
Excellent, just two error message fixes.

File: src/express/outgoing/manager.ts (hunk 17 above)
Comment (yarinBenisty) [reply src/express/outgoing/manager.ts:16]: done

File: src/express/outgoing/manager.ts (hunk 18 above)
Comment (yarinBenisty) [reply src/express/outgoing/manager.ts:68]: done

**Review by MichaelSimkin — APPROVED** [review]

**Pull request merged**